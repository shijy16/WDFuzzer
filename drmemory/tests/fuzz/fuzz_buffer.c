/* **************************************************************
 * Copyright (c) 2015 Google, Inc.  All rights reserved.
 * **************************************************************/

/*
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Google, Inc. nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */

/* Test of the Dr. Memory Fuzz Testing Feature */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef WINDOWS
# define EXPORT __declspec(dllexport)
#else
# define EXPORT
#endif

#define ELEMENT_SIZE (sizeof(unsigned int))

typedef unsigned int uint;

typedef enum _bool {
    false,
    true
} bool;

typedef struct _deliberate_errors_t {
    bool overread;
    bool underread;
    bool overwrite;
    bool underwrite;
    bool leak;
    bool uninit;
    uint fuzz_iteration;
} deliberate_errors_t;

static deliberate_errors_t deliberate_errors;

/* print the contents of the buffer as unsigned integers */
EXPORT void
repeatme(uint *buffer, size_t size)
{
    uint i, elements = (size / ELEMENT_SIZE);

    if (deliberate_errors.uninit) {
        uint val;
        for (i = 0; i < elements; i++) {
            /* buffer value should have at most one bit set */
            val = buffer[i] & (buffer[i] - 1);
            if (val != 0)
                printf("Error: mutator flipped too many bits: %u!\n", val);
        }
    } else {
        printf("Buffer:");
        for (i = 0; i < elements; i++)
            printf(" 0x%08x", buffer[i]);
        printf("\n");
    }

    if ((++deliberate_errors.fuzz_iteration % 2) == 0) {
        if (deliberate_errors.overread)
            printf("over-read: %d\n", buffer[elements + 1]);
        if (deliberate_errors.underread)
            printf("under-read: %d\n", *(buffer - 1));
        if (deliberate_errors.overwrite) {
            buffer[elements] = 7;
        }
        if (deliberate_errors.underwrite)
            *(buffer - 1) = 7;
    }
}

int
main(int argc, char **argv)
{
    uint i, j, elements = 4, size = elements * ELEMENT_SIZE, *buffer = malloc(size);
    bool testing_uninit = false;

    if (argc > 1 && strcmp(argv[1], "initialize") == 0) {
        for (i = 0; i < elements; i++)
            buffer[i] = (i + 1);
    } else {
        testing_uninit = true;
        deliberate_errors.uninit = true;
    }


    if (argc > 2) {
        if (strcmp(argv[2], "overread") == 0)
            deliberate_errors.overread = true;
        else if (strcmp(argv[2], "underread") == 0)
            deliberate_errors.underread = true;
        else if (strcmp(argv[2], "overwrite") == 0)
            deliberate_errors.overwrite = true;
        else if (strcmp(argv[2], "underwrite") == 0)
            deliberate_errors.underwrite = true;
        else if (strcmp(argv[2], "leak") == 0)
            deliberate_errors.leak = true;
    }

    repeatme(buffer, size);

    deliberate_errors.uninit = false;
    if (!deliberate_errors.leak && !testing_uninit/*verify that init is detected */)
        free(buffer);

    for (j = 1; j < 3; j++) {
        elements--;
        size = elements * ELEMENT_SIZE;
        if (!testing_uninit) /* verify that init is maintained for the whole fuzz pass */
            buffer = malloc(size);
        for (i = 0; i < elements; i++)
            buffer[i] = (i + j + 1);
        repeatme(buffer, size);
        if (!testing_uninit) /* reusing the buffer for this test */
            free(buffer);
    }

    if (testing_uninit)
        free(buffer); /* now free for the one test that kept the original buffer */

    printf("done\n");
    return 0;
}
