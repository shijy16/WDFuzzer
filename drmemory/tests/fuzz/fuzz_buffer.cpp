/* **************************************************************
 * Copyright (c) 2015 Google, Inc.  All rights reserved.
 * **************************************************************/

/*
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Google, Inc. nor the names of its contributors may be
 *   used to endorse or promote products derived from this software without
 *   specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */

/* Test of the Dr. Memory Fuzz Testing Feature */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef WINDOWS
# define EXPORT __declspec(dllexport)
#else
# define EXPORT
#endif

#define BUFFER_ELEMENTS 4
#define ELEMENT_SIZE (sizeof(uint))

typedef unsigned int uint;

class BufferPrinter;

class DeliberateErrors {
    friend class BufferPrinter;

    public:
        DeliberateErrors(const char *deliberate_error);

    private:
        bool uninit;
        bool overread;
        bool underread;
        bool overwrite;
        bool underwrite;
        bool leak;
        uint fuzz_iteration;
};

class BufferPrinter {

    public:
        /* print the contents of the buffer as unsigned integers */
        EXPORT void
        repeatme(uint *buffer, size_t size);

        BufferPrinter(const char *deliberate_error);

        ~BufferPrinter();

    private:
        DeliberateErrors *deliberate_errors;
};

BufferPrinter::BufferPrinter(const char *deliberate_error)
{
    deliberate_errors = new DeliberateErrors(deliberate_error);
}

BufferPrinter::~BufferPrinter()
{
    if (!deliberate_errors->leak)
        delete deliberate_errors;
}

EXPORT void
BufferPrinter::repeatme(uint *buffer, size_t size)
{
    uint i = 0, elements = (size / ELEMENT_SIZE);

    if (deliberate_errors->uninit) {
        uint val;
        for (i = 0; i < elements; i++) {
            /* buffer value should have at most one bit set */
            val = buffer[i] & (buffer[i] - 1);
            if (val != 0)
                printf("Error: mutator flipped too many bits: %u!\n", val);
        }
    } else {
        printf("Buffer:");
        for (i = 0; i < elements; i++)
            printf(" 0x%08x", buffer[i]);
        printf("\n");
    }

    if ((++deliberate_errors->fuzz_iteration % 2) == 0) {
        if (deliberate_errors->overread)
            printf("over-read: %d\n", buffer[elements + 1]);
        if (deliberate_errors->underread)
            printf("under-read: %d\n", *(buffer - 1));
        if (deliberate_errors->overwrite)
            buffer[elements] = 7;
        if (deliberate_errors->underwrite)
            *(buffer - 1) = 7;
    }
}

DeliberateErrors::DeliberateErrors(const char *deliberate_error)
{
    this->uninit = (strcmp(deliberate_error, "uninit") == 0);
    this->overread = (strcmp(deliberate_error, "overread") == 0);
    this->underread = (strcmp(deliberate_error, "underread") == 0);
    this->overwrite = (strcmp(deliberate_error, "overwrite") == 0);
    this->underwrite = (strcmp(deliberate_error, "underwrite") == 0);
    this->leak = (strcmp(deliberate_error, "leak") == 0);
    this->fuzz_iteration = 0;
}

int
main(int argc, char **argv)
{
    uint i, size = BUFFER_ELEMENTS * ELEMENT_SIZE, *buffer = new uint[BUFFER_ELEMENTS];
    /* assuming argv[1] must be "initialize" if exists */
    const char *deliberate_error = (argc > 2 ? argv[2] : (argc > 1 ? "" : "uninit"));
    BufferPrinter bp(deliberate_error);

    if (argc > 1) {
        /* argv[1] must be "initialize" */
        if (strcmp(argv[1], "initialize") != 0)
            return 1;
        for (i = 0; i < BUFFER_ELEMENTS; i++)
            buffer[i] = (i + 1);
    }

    bp.repeatme(buffer, size);

    delete [] buffer;

    printf("done\n");
    return 0;
}
