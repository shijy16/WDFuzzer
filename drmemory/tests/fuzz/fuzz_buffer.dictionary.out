# **********************************************************
# Copyright (c) 2015-2016 Google, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
Buffer: 0x666f6f77 0x00000020 0x00000003 0x00000004
Buffer: 0x656b6f74 0x2032346e 0x00000003 0x00000004
Buffer: 0x66666964 0x6e657265 0x6f745f74 0x206e656b
Buffer: 0x0020efcd 0x00000002 0x00000003 0x00000004
Buffer: 0x65746e69 0x6c616e72 0x6f757122 0x22736574
Buffer: 0x6b636162 0x616c735c 0x20236873 0x00000004
Buffer: 0x666f6f77 0x00000020 0x00000004
Buffer: 0x656b6f74 0x2032346e 0x00000004
Buffer: 0x66666964 0x6e657265 0x6f745f74
Buffer: 0x0020efcd 0x00000003 0x00000004
Buffer: 0x65746e69 0x6c616e72 0x6f757122
Buffer: 0x6b636162 0x616c735c 0x20236873
Buffer: 0x666f6f77 0x00000020
Buffer: 0x656b6f74 0x2032346e
Buffer: 0x66666964 0x6e657265
Buffer: 0x0020efcd 0x00000004
Buffer: 0x65746e69 0x6c616e72
Buffer: 0x6b636162 0x616c735c
done
~~Dr.M~~ NO ERRORS FOUND:
~~Dr.M~~       0 unique,     0 total unaddressable access(es)
~~Dr.M~~       0 unique,     0 total invalid heap argument(s)
~~Dr.M~~       0 unique,     0 total warning(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of leak(s)
~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
